Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs

```diff
@@ -0,0 +1,254 @@
+                var guardian = _guardianRepository.GetGuardian(id);
+                var guardianElo = _guardianEloRepository.GetGuardianElo(id).NewElo;
+                var guardianEff = _guardianEfficiencyRepository.GetGuardianEfficiency(id).NewEfficiency;
+                
+                var snapshot = new GuardianSnapshot()
+                {
+                    Guardian = guardian,
+                    GuardianEfficiency = guardianEff,
+                    GuardianElo = guardianElo
+                };
+                
+                guardians.Add(snapshot);
+            }
+            
+            var teams = RandomizeTeams(guardians, teamSize);
+
+            teams = Anneal(teams);
+            
+            return teams;
+        }
+
+        private IEnumerable<MatchMadeTeam> RandomizeTeams(IEnumerable<GuardianSnapshot> guardians, int teamSize)
+        {
+            var numberOfTeams = guardians.Count() / teamSize;
+            var guardianArray = ShuffleGuardiansList(guardians);
+            
+            var teams = new List<MatchMadeTeam>();
+            for (var i = 0; i < numberOfTeams; i++)
+            {
+                var team = new MatchMadeTeam();
+                var teamGuardians = new List<GuardianSnapshot>();
+                var teamElo = 0.0;
+                var teamEfficiency = 0.0;
+                
+                for (var j = i * teamSize; j < i * teamSize + teamSize; j++)
+                {
+                    teamGuardians.Add(guardianArray[j]);
+                    teamElo += guardianArray[j].GuardianElo;
+                    teamEfficiency += guardianArray[j].GuardianEfficiency;
+                }
+
+                team.Guardians = teamGuardians;
+                team.TeamElo = teamElo / teamSize;
+                team.TeamEfficiency = teamEfficiency / teamSize;
+                
+                teams.Add(team);
+            }
+
+            return teams;
+        }
+
+        private IEnumerable<MatchMadeTeam> Anneal(IEnumerable<MatchMadeTeam> teams)
+        {
+            var temperature = 400.0;
+
+            var iterations = 0;
+            while (temperature > _epsilon)
+            {
+                iterations++;
+
+                var (teamIndex1, teamIndex2) = GetRandomTeamIndexesForSwap(teams);
+                var nextTeamConfiguration = SwapTwoPlayers(teams, teamIndex1, teamIndex2);
+                
+                var teamElos = GetTeamElos(teams).ToArray();
+                var newTeamElos = GetTeamElos(nextTeamConfiguration).ToArray();
+                
+                var startingDelta = CalculateDelta(teamElos[teamIndex1], teamElos[teamIndex2]);
+                var newConfigurationDelta = CalculateDelta(newTeamElos[teamIndex1], newTeamElos[teamIndex2]);
+
+                var delta = newConfigurationDelta - startingDelta;
+                if (delta < 0)
+                {
+                    teams = KeepNextConfiguration(nextTeamConfiguration);
+                }
```

Comment:
Without doing the annealing part below you're essentially doing a random hill climb, which will find a local maxima but might miss a global maxima. I'm not an expert in matchmaking algorithms but I think having a less accurate global maxima is the preferred solution.